Updates to deepTau@HLT

lwezenbe opened the pull request:
**Changes:**

- Added possibility to check if tau to be processed is in a certain collection. Use case: Make sure deeptau only runs for taus that are matched to L1 in order to reduce processing time. In case match is required and not met, a deeptau output of -1 is given, which will fail the following filters
- Removed deepTauAtHLT.py customisation file and replaced with two customisation files that simulate a full ditau path with deepTau for i.e. timing measurements. These can be run with the full menu. Module names are also adapted to not cause unwanted overlap with other paths.
- Implemented proper deepTau working points. These are a function of tau pt. PFTauSelectorDefinition.h has been updated to allow for filtering on deeptau output.

**Testing:**

- DeepTau modules have been run both for offline and online (with the two customisation functions) and run smoothly without unexpected surprises in the output. 
- The following code has been run to perform tests, which all succeeded:
```
scram b distclean 
git cms-checkdeps -a -A
scram b -j 8
scram b runtests
runTheMatrix.py -l limited -i all --ibeos
```

**NOTE**
After integration into the cms-tau-pog fork, it would be good if we could hold off on merging this centrally until one of the developers who is currently on holiday confirms there are no other changes required except for the ones in this PR

kandrosov commented:
I do not understand why `taus_to_compare` is added. I think it would be better to have the producer work with the selected tau collection, instead of doing the match.

mbluj commented:
> I do not understand why `taus_to_compare` is added. I think it would be better to have the producer work with the selected tau collection, instead of doing the match.

I agree with @kandrosov that it will be better to run deepTau on top of a final tau collection than perform the matching. However, I recall (to be checked) that L1-matched tau collection at HLT is simplified, i.e. it consists only basic information on taus (4-momentum, etc.). If it is the case it should be considered to rethink how the collection is built - dR-based matching inside deepTau does not look to be the optimal solution.

kandrosov commented:
In my opinion, it would be better to change the collection returned by the L1 matching module to a vector of references to the original taus instead of copies. And modify the DeepTau producer in order to be able to work on both types of inputs.

mbluj commented:
> In my opinion, it would be better to change the collection returned by the L1 matching module to a vector of references to the original taus instead of copies. And modify the DeepTau producer in order to be able to work on both types of inputs.

+1
I like this idea. However, it (at least modification of the L1 matching module) looks be beyond scope of this PR => I propose to remove the matching here, check how the L1 matching module should be modified and then open a new PR (or two PRs one for L1-mathing other enabling a new type of input collection (I think it it should be RefVector or so)).

lwezenbe commented:
Agreed, I tried to think of why it was done like this in the first place but I can not find any good reason. I have already removed the reference filter from the code and will take a look at your suggestion next week

kandrosov commented on RecoTauTag/RecoTau/src/DeepTauBase.cc:
```diff
@@ -26,11 +26,11 @@ namespace deep_tau {
       static const std::string prefix =
           "[&](double *x, double *p) { const int decayMode = p[0];"
-          "const double pt = p[1]; const double eta = p[2];";
+          "const double pt = p[1]; const double eta = p[2]; return ";
```
Originally `return ` was not added here to keep a possibility of having more complicated cut expressions like, for example, this one https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py#L745-L748

I would prefer to keep such flexibility. But to improve user-friendliness for not so complicated cut expressions, you can modify the code in the following way:
```c++
std::string fn_str = prefix;
if(cut_str.find("return") == std::string::npos)
  fn_str += " return " + cut_str + ";}";
else
  fn_str += cut_str + "}";
```

mbluj replied:
A followup needed? Or it is resolved?

lwezenbe replied:
I will change it to be in the working point definition instead of here, I will change this now

mbluj commented:
> Agreed, I tried to think of why it was done like this in the first place but I can not find any good reason. I have already removed the reference filter from the code and will take a look at your suggestion next week

I've checked this: indeed L1-matched PFTaus are "trivial" in the sense that they contain only p4 and vertex, see [L1THLTTauMatching.cc#L50](https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/HLTProducers/src/L1THLTTauMatching.cc#L50). As a PFTau vertex the vertex of its leading charged particle is set. This, as far as I remember, is intentional - it was to better define dz of di-tau. Anyway, one can try to copy original taus selected taus instead of building their "trivial" copies and then set the new vertex if indeed needed.

lwezenbe commented:
I have been working on this on and off for the past few weeks and although making the output of the L1 matcher a refVector is easy to do and deepTauProducer can already handle this sort of input with no modification, I have been running into a larger problem with the next step in the process where we filter on the number of taus that passed deepTau.

When this step is reached:
https://github.com/lwezenbe/cmssw/blob/CMSSW_12_0_X_deepTau/RecoTauTag/HLTProducers/python/deepTauAtHLT_35GeV.py#L180-L195
there are some issues when running the deepTauProducer with the RefVector input.
Although it is not shown here, a src is required with the tau collection. 

Entering the RefVector makes it crash with the following error:
```
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<reco::PFTau>
```

When trying to solve this, I have the feeling I have little to work with since loading the collection goes through the following code: https://github.com/cms-sw/cmssw/blob/master/CommonTools/UtilAlgos/interface/ObjectSelector.h#L63-L75

When entering the original tau collection, it will also crash because it tries to access information that is not there since the taus that did not pass the L1 filter are not saved in the DeepTauProducer.

I will continue to work on this but decided to share the status in case I am missing something obvious that you can already see.

swozniewski commented:
Hi @lwezenbe, can you post the commands you are running? I'd like to reproduce and investigate a bit myself unless you made further progress in the meantime.

lwezenbe commented:
Yes, sure, I will make a different PR for this update and write there what command I am running. I thought I had a solution by using edm::View in that module, but it turns out I had to change the output collection to one that does not work with the next module in the path for that so I will be trying different things today

lwezenbe commented:
Since this is a different development than what was envisaged here, I have opened a PR for it:
https://github.com/cms-tau-pog/cmssw/pull/152

I have added the commands I ran there.

lwezenbe commented:
Just as an update, I will adapt the customisation files for the 35GeV and 30 GeV ditau paths with deepTau here after the other PR has been merged.

Additionally, Anna has found some potential issue with some input variables that we are now studying so we would also implement the fix here if one is needed so that we do not need to keep making PRs to optimize deepTau@HLT

mbluj commented:
Hi,
it concerns comment from another PR #152: it seems to not be necessary to revert changes in customisation file `RecoTauTag/HLTProducers/python/deepTauAtHLT.py` as it was removed in this PR (replaced with two customisation files), so disregard my comments in #152. On the other hand merge of #152 causes a merge conflict in this PR - please rebase your development branch to solve the conflict.
Is there still anything missing in this PR (except solving the merge conflict)? If not, I would like to proceed with final review and merging, and then proceed with integration to the official CMSSW.

mbluj commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -1149,6 +1149,7 @@ class DeepTauId : public deep_tau::DeepTauBase {
     desc.add<edm::InputTag>("electrons", edm::InputTag("slimmedElectrons"));
     desc.add<edm::InputTag>("muons", edm::InputTag("slimmedMuons"));
     desc.add<edm::InputTag>("taus", edm::InputTag("slimmedTaus"));
+    desc.add<edm::InputTag>("taus_to_compare", edm::InputTag("default"));
```
It is not needed any longer, right? If so, please remove.

mbluj commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -1532,6 +1533,13 @@ class DeepTauId : public deep_tau::DeepTauBase {
                 << "invalid prediction = " << pred << " for tau_index = " << tau_index << ", pred_index = " << k;
           predictions.matrix<float>()(tau_index, k) = pred;
         }
+      } else {
```
This `else` cause is to set "default" output values in case tau does not pass a preselection? What was set w/o this cause, zero? I suppose that it is not a problem with deepTau run at miniAOD as there is not any preselection, right?

lwezenbe replied:
Indeed, I noticed that the deepTau was actually getting random values, so I set this as default, the way it is set here is that it the deeptau vs jets is set to -1. And as you indicated, the deepTau at miniAOD does not use preselection, this was added in the development of DeepTau@HLT

mbluj commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -1532,6 +1533,13 @@ class DeepTauId : public deep_tau::DeepTauBase {
         }
+      } else {
+        for (int k = 0; k < deep_tau::NumberOfOutputs; ++k) {
+          if (k == 2)
```
Why `k==2` is special?

lwezenbe replied:
It's been hardcoded this way initially because k == 2 is the index for the tau in the [list of outputs](https://github.com/cms-tau-pog/cmssw/blob/ad3236873b5fe6395b89de6eb3981357c1767d96/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L1106). With the rest set to 2, it ended up as -1. in [this calculation](https://github.com/cms-tau-pog/cmssw/blob/ad3236873b5fe6395b89de6eb3981357c1767d96/RecoTauTag/RecoTau/src/DeepTauBase.cc#L68-L69). If you have a preference for a different way of setting defaults, this can be changed

mbluj replied:
The default can change in case `deep_tau::NumberOfOutputs` has different value. But maybe it is not a problem. Anyway, I would add to the code a comment explaining this for reference.

mbluj commented on RecoTauTag/RecoTau/plugins/PFTauSelectorDefinition.h:
```diff
@@ -126,6 +126,8 @@ struct PFTauSelectorDefinition {
                   << "PFTauSelector: Requested working point '" << disc.rawLabels[i] << "' not found!\n";
           }
+        } else if (psetsFromProvenance.exists("VSjetWP")) {
+          ;
```
Is this empty block really needed? It is to correctly throw an exception below?

lwezenbe replied:
If this block isnt there, it will go the the else statement since the DeepTauId does not have an IDdefinitions or rawValues but only the workingPointsVsJets and so on. So without this block it will always throw an exception. Since in the HLT no rawValues are used but only a few set working points, it does not need to do anything here, which is why I left it empty. To be honest, I was not completely sure what the best way to deal with this was, so I am open to put something else here if you have better suggestions. My reasoning was that the less it has to do at HLT , the better.

mbluj replied:
It is OK. But I would add a comment for future developers.

mbluj reviewed (changes requested):
I went trough the changes and I have just a few small comments/questions - please check.

lwezenbe commented:
I am just going through the code, updating the customisation functions to work with the updated L1 matcher and I just remembered that when trying to access the module in confDB when making the first skeleton of the updated menu, I did not have access to the working point input parameters. Coincidentally they are also the only ones that do not have a default in the fillDescriptions so I am wondering if it is connected. Would it be ok to put an empty vector as a default or is there a reason why these do not have a default?

mbluj commented:
> I am just going through the code, updating the customisation functions to work with the updated L1 matcher and I just remembered that when trying to access the module in confDB when making the first skeleton of the updated menu, I did not have access to the working point input parameters. Coincidentally they are also the only ones that do not have a default in the fillDescriptions so I am wondering if it is connected. Would it be ok to put an empty vector as a default or is there a reason why these do not have a default?

I think it might be connected - confDB parses module's cfi (given explicitely or via fillDescriptions) to build a template configuration for the module. Therefore, if a parameter is not specified in the cfi it is not defined in confDB. It means that some more or less dummy WPs should be added and have a correct data-type.
@kandrosov could you comment if adding an empty vector is fine, please? Or one should put some values, e.g. allowing to pass all events (-1)? (Empty vector seems to be a better solution as will not produce useless outputs).

mbluj commented:
Follow up clarification: cfi parsing to ConfDB is not automatic - it can be done on request when the code is integrated to an official release. In the pasts it was executed by Martin Gruenewald (maybe he is still in charge?)

lwezenbe commented:
I think all comments are answered except for the point on the default values for the WP in the fillDescriptions. I wanted to run all tests again before the end of the day but I encountered some technical issues that prevented me from completing them by now so I unfortunately will not be able to do those anymore today.

mbluj commented:
> I think all comments are answered except for the point on the default values for the WP in the fillDescriptions. I wanted to run all tests again before the end of the day but I encountered some technical issues that prevented me from completing them by now so I unfortunately will not be able to do those anymore today.

Thank you! But, if I am not mistaken Konstantin's comment on cut parser (in RecoTauTag/RecoTau/src/DeepTauBase.cc) still should be addressed. Please also run code-checks and code-format before matrix tests.

lwezenbe commented:
Ah sorry, I indeed missed that. I'm facing a bit of an issue with running the tests. It seems that the base branch that this PR is merging into is in a state in between CMSSW_12_0_0_pre5 and CMSSW_12_0_0_pre6. I had to reinstall my environment due to technical issues with it but now I am getting a ton of additional commits that have nothing to do with these changes and I keep hitting a broken pipe when compiling due to the amount of changes that are now present. Could you rebase this branch to CMSSW_12_0_0? I was also going to do the same to my base branch that I use to merge here but I think if only I do it, this PR will contain a ton of irrelevant commits as well.

lwezenbe commented:
Actually, now that I am looking deeper into it, it seems that the commits in the cms-tau-pog version on August 3rd are different from the one in the central CMSSW, I'm not sure what is going on there.

mbluj commented:
Could it be mix with some stuff for 12_1_X? I will try to rebase to this release series as our target is there. In case of need we will try to backport to 12_0_X later. I will let you know

mbluj commented:
@lwezenbe I have just rebased the  `cms-tau-pog:CMSSW_12_0_X_tau-pog_DeepTauAtHLT` branch to CMSSW_12_1_0_pre4 which is the most recent development release. Please do the same with your development branch like this ([see here for details](https://cms-sw.github.io/tutorial-resolve-conflicts.html)):
```
cmsrel CMSSW_12_1_0_pre4
cd CMSSW_12_1_0_pre4/src
cmsenv
git cms-init
git cms-rebase-topic lwezenbe:CMSSW_12_0_X_deepTau
# solve possible conflicts and then git add; git rebase --continue
# check if it compiles with scram b
# push rebased branch to your repository
git push -f my-cmssw CMSSW_12_0_X_deepTau # -f is mandatory as rebasing rewrites the history
```
I hope it will work and solve the issues.

mbluj commented:
P.S. I think that you can rebase your branch on top of this branch when you execute the following command
```
 git cms-rebase-topic cms-tau-pog:CMSSW_12_0_X_tau-pog_DeepTauAtHLT
```
before rebasing your branch. Rebasing of `cms-tau-pog:CMSSW_12_0_X_tau-pog_DeepTauAtHL` will be "trivial" as it is already rebased, but it will checkout the branch and then use as the base branch.

lwezenbe commented:
Worked like a charm, I will update the final things and run  all tests so that it is ready by the end of the day

mbluj commented:
> Worked like a charm, I will update the final things and run all tests so that it is ready by the end of the day

OK, great!

mbluj commented:
Please update also description of the PR as e.g. some of changes were finally not implemented (matching), others are not described (defaults when a tau does not pass a preselection), note about holding integration with official CMSSW seems outdated, etc.

lwezenbe commented:
I have now added Konstantin's request, I believe I have added all changes to the descriptions now and the last tests just finished running, there were 0 failed tests

mbluj commented:
Thank you! It is now merged. I will proceed with PR to official CMSSW.
